File: frontend/vizr/vizr/symlog.py
"""
Author: Sobolev Y. A.
Description: Provides scientific number handling and symmetric logarithmic scaling for visualization.
"""

import numpy as np

# we define a very small threshold for the linear region around zero.
# in a standard logarithmic scale, log(0) is undefined and log(negative) is complex.
# symmetric log (symlog) solves this by being linear around zero and logarithmic elsewhere.
# LINTHRESH is the range (-LINTHRESH, LINTHRESH) where the plot is linear.
# LOG_LINTHRESH is simply log10(LINTHRESH) precomputed for speed.
LINTHRESH = 1e-50
LOG_LINTHRESH = -50.0


class Scientific:
    """
    A custom arbitrary-precision floating point number representation.
    Not to lose precision when loading from Parquet.
    """

    def __init__(self, mantissa: float, exponent: int):
        """
        Initializes a Scientific number.

        Inputs:
            mantissa (float): The significant digits of the number.
            exponent (int): The power of 10 to multiply the mantissa by.
        """
        self.mantissa = mantissa
        self.exponent = exponent

    @classmethod
    def from_str(cls, s: str) -> "Scientific":
        """
        Parses a scientific string (e.g., "1.23e-100") into a Scientific object.
        We expect strings because standard JSON or Parquet floats often lack the bits
        needed for exponents like -10000.

        Inputs:
            s (str): The string representation of the number (e.g., "1.2e-5").

        Outputs:
            Scientific: The parsed Scientific object.
        """
        if not s:
            return cls(0.0, 0)
        
        # standardize the string to handle 'e' or 'E'.
        s = s.lower()
        if "e" in s:
            parts = s.split("e")
            mantissa = float(parts[0])
            exponent = int(parts[1])
            return cls(mantissa, exponent)
        else:
            # if there is no exponent part, it is just a regular number with exponent 0.
            return cls(float(s), 0)

    def approx_f64(self) -> float:
        """
        Converts to a standard Python float (64-bit).
        Warning: May overflow/underflow if the exponent is wild. Obviously.
        This is mostly used when we know the value is within displayable range
        or when we accept the loss of precision for a pixel coordinate.

        Outputs:
            float: The 64-bit floating point approximation.
        """
        return self.mantissa * (10.0**self.exponent)

    def symlog(self) -> float:
        """
        Computes the Symmetric Logarithmic value.
        Useful for plotting data that spans many orders of magnitude and includes zero/negative values.
        
        The transformation function is roughly:
        y = sign(x) * (log10(|x|) - log10(threshold)) for |x| > threshold
        y = sign(x) * linear_interpolation(...) for |x| <= threshold

        Outputs:
            float: The symmetric logarithm of the value.
        """
        if self.mantissa == 0.0:
            return 0.0

        sign = np.sign(self.mantissa)
        abs_mantissa = abs(self.mantissa)

        # calculate the log10 magnitude using the separate exponent to avoid underflow
        # before taking the log.
        # log10(m * 10^e) = log10(m) + e
        val_log10 = np.log10(abs_mantissa) + self.exponent

        # how far are we from the linear threshold?
        magnitude_diff = val_log10 - LOG_LINTHRESH

        # if we are significantly larger than the threshold, we use the log behavior.
        # we use 16.0 as a somewhat arbitrary cutoff where we switch to pure log behavior.
        if magnitude_diff > 16.0:
            return sign * magnitude_diff
        else:
            # in the transition region or near zero, we use a function that behaves linearly
            # to avoid the singularity at zero.
            return sign * np.log10(1.0 + abs(self.approx_f64()) / LINTHRESH)

    def format(self) -> str:
        """
        Formats the number nicely for display.
        We want short strings for tooltips, so we strip unnecessary zeros.

        Outputs:
            str: A string representation suitable for UI display.
        """
        if self.mantissa == 0.0:
            return "0"

        sign_str = "-" if self.mantissa < 0.0 else ""
        abs_mantissa = abs(self.mantissa)

        # if the number is very large or very small, force scientific notation.
        if self.exponent < -2 or self.exponent > 3:
            return f"{sign_str}{abs_mantissa:.1f}e{self.exponent:.0f}"
        else:
            # otherwise, try to show it as a normal decimal number.
            real_val = abs_mantissa * (10.0**self.exponent)
            # remove trailing zeros and decimal point if not needed
            return f"{sign_str}{real_val:.6f}".rstrip("0").rstrip(".")


def vectorized_approx_f64(mantissa: np.ndarray, exponent: np.ndarray) -> np.ndarray:
    """
    Vectorized version of approx_f64 using numpy arrays.
    Essential for plotting performance when handling thousands of points.

    Inputs:
        mantissa (np.ndarray): Array of mantissa values.
        exponent (np.ndarray): Array of exponent values.

    Outputs:
        np.ndarray: Array of approximate float64 values.
    """
    return mantissa * (10.0**exponent)


def vectorized_symlog(mantissa: np.ndarray, exponent: np.ndarray) -> np.ndarray:
    """
    Vectorized Symmetric Log transformation for arrays of mantissas/exponents.
    High performance version for plotting large datasets.
    
    This function replicates the logic of Scientific.symlog but operates on entire
    numpy arrays at once using masks.

    Inputs:
        mantissa (np.ndarray): Array of mantissas.
        exponent (np.ndarray): Array of exponents.

    Outputs:
        np.ndarray: Array of transformed values on the SymLog scale.
    """
    # handle zeros first. The result for 0 is 0.
    result = np.zeros_like(mantissa)
    non_zero = mantissa != 0

    if not np.any(non_zero):
        return result

    # we only process non-zero elements to avoid log(0) warnings.
    m = mantissa[non_zero]
    e = exponent[non_zero]

    sign = np.sign(m)
    abs_mantissa = np.abs(m)

    # log10(|x|) = log10(|m|) + e
    val_log10 = np.log10(abs_mantissa) + e

    magnitude_diff = val_log10 - LOG_LINTHRESH

    # we treat numbers much larger than the threshold differently than those close to zero.
    huge_mask = magnitude_diff > 16.0
    small_mask = ~huge_mask

    # prepare output container for the non-zero elements
    out_vals = np.zeros_like(m)

    # for huge numbers, the linear term 1.0 is negligible, so we just use the log difference.
    if np.any(huge_mask):
        out_vals[huge_mask] = sign[huge_mask] * magnitude_diff[huge_mask]

    # for small numbers, we compute the full formula to ensure smooth transition through zero.
    if np.any(small_mask):
        approx_val = m[small_mask] * (10.0 ** e[small_mask])
        out_vals[small_mask] = sign[small_mask] * np.log10(
            1.0 + np.abs(approx_val) / LINTHRESH
        )

    # scatter the computed values back into the result array
    result[non_zero] = out_vals

    return result


def symlog_formatter(val: float) -> str:
    """
    Formats a value on the SymLog axis for humans to read.
    Reverses the SymLog math to show actual values.
    
    The axis ticks are in the transformed 'symlog' space, so we need to
    invert the transformation to show the user the real number (e.g., 1e-50).

    Inputs:
        val (float): The value in SymLog space (from the plot axis).

    Outputs:
        str: The human-readable string representation of the original value.
    """
    if val == 0.0:
        return "0"
    # if we are right on the threshold, we just show 1 (scaled) or similar basic value.
    elif abs(val + LOG_LINTHRESH) < 0.00001:
        return "1"

    sign_str = "-" if val < 0.0 else ""
    abs_plot_y = abs(val)

    # invert the log transformation:
    # plot_y = log10(x) - log10(threshold)
    # log10(x) = plot_y + log10(threshold)
    target_log10 = LOG_LINTHRESH + abs_plot_y

    exponent = np.floor(target_log10)
    fractional = target_log10 - exponent
    mantissa = 10.0**fractional

    # choose format based on magnitude
    if exponent < -2.0 or exponent > 3.0:
        return f"{sign_str}{mantissa:.1f}e{exponent:.0f}"
    else:
        real_val = mantissa * (10.0 ** int(exponent))
        return f"{sign_str}{real_val:.4f}".rstrip("0").rstrip(".")

File: shanks-unified/src/plot/symlog.rs
//! Symlog scale for plotting values with large dynamic range.
//!
//! Ported from old-vizr/src/symlog.rs

use serde::{Deserialize, Serialize};

/// A value in scientific notation: mantissa × 10^exponent.
#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize)]
pub struct Scientific(pub f64, pub i32);

impl Scientific {
    /// Convert to approximate f64 value.
    /// May lose precision for very large/small values.
    pub fn approx_f64(&self) -> f64 {
        self.0 * 10f64.powi(self.1)
    }

    /// Create from f64 value.
    pub fn from_f64(val: f64) -> Self {
        if val == 0.0 {
            return Scientific(0.0, 0);
        }

        let abs_val = val.abs();
        let sign = val.signum();
        let log10 = abs_val.log10();
        let exponent = log10.floor() as i32;
        let mantissa = sign * abs_val / 10f64.powi(exponent);

        Scientific(mantissa, exponent)
    }

    /// Apply symlog transformation for plotting.
    ///
    /// Symlog provides a smooth transition between linear and logarithmic scales:
    /// - Near zero: linear (preserves sign and small differences)
    /// - Far from zero: logarithmic (handles large dynamic range)
    pub fn symlog(&self, log_linthresh: f64) -> f64 {
        let linthresh = 10f64.powf(log_linthresh);
        let mantissa = self.0;
        let exponent = self.1;

        // Handle zero
        if mantissa == 0.0 {
            return 0.0;
        }

        let sign = mantissa.signum();
        let abs_mantissa = mantissa.abs();

        // log10(|x|)
        let val_log10 = abs_mantissa.log10() + exponent as f64;

        // Compare magnitude vs threshold
        // If the value is more than 16 orders of magnitude larger than threshold,
        // the "+ 1" in symlog formula becomes irrelevant due to f64 precision limits.
        let magnitude_diff = val_log10 - log_linthresh;

        if magnitude_diff > 16.0 {
            // Huge numbers: log approximation
            // Formula: log10(|x|) - log10(L)
            // This preserves precision for massive numbers (e.g. 1e100) avoiding overflow.
            sign * magnitude_diff
        } else {
            // Small/transition numbers: exact math
            // Formula: log10(1 + |x|/L)
            // Near threshold, the "+ 1" creates the smooth curve.
            sign * (1.0 + self.approx_f64().abs() / linthresh).log10()
        }
    }

    /// Format as a human-readable string.
    pub fn format(&self) -> String {
        if self.0 == 0.0 {
            return "0".to_string();
        }

        let mantissa = self.0;
        let exponent = self.1;
        let sign_str = if mantissa < 0.0 { "-" } else { "" };
        let abs_mantissa = mantissa.abs();

        // Use scientific notation for very small or very large numbers
        if exponent < -2 || exponent > 3 {
            format!("{}{:.1}e{:.0}", sign_str, abs_mantissa, exponent)
        } else {
            // For numbers like 0.5, 0.01, 10.0
            let real_val = abs_mantissa * 10f64.powi(exponent);
            format!("{}{:.20}", sign_str, real_val)
        }
    }
}

impl From<f64> for Scientific {
    fn from(val: f64) -> Self {
        Self::from_f64(val)
    }
}

impl From<Scientific> for f64 {
    fn from(val: Scientific) -> Self {
        val.approx_f64()
    }
}

/// Convert symlog-transformed value back to a formatted string.
///
/// This is used for axis labels on symlog-scaled plots.
pub fn symlog_formatter(val: f64, log_linthresh: f64) -> String {
    if val == 0.0 {
        return "0".to_string();
    } else if (val + log_linthresh).abs() < 0.00001 {
        return "1".to_string();
    }

    let sign_str = if val < 0.0 { "-" } else { "" };
    let abs_plot_y = val.abs();

    // Inverse transform: |x| = L * (10^|y| - 1)
    // In log region: |x| ~= L * 10^|y|
    // So: log10(|x|) = log10(L) + |y|
    let target_log10 = log_linthresh + abs_plot_y;

    // Reconstruct scientific notation
    let exponent = target_log10.floor();
    let fractional = target_log10 - exponent;
    let mantissa = 10f64.powf(fractional);

    // Format
    if exponent < -2.0 || exponent > 3.0 {
        format!("{}{:.1}e{:.0}", sign_str, mantissa, exponent)
    } else {
        let real_val = mantissa * 10f64.powi(exponent as i32);
        format!("{}{:.4}", sign_str, real_val)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_scientific_from_f64() {
        let s = Scientific::from_f64(123.456);
        assert!((s.0 - 1.23456).abs() < 0.0001);
        assert_eq!(s.1, 2);
    }

    #[test]
    fn test_scientific_approx_f64() {
        let s = Scientific(1.23, 5);
        assert!((s.approx_f64() - 123000.0).abs() < 0.1);
    }

    #[test]
    fn test_symlog_zero() {
        let s = Scientific(0.0, 0);
        assert_eq!(s.symlog(-50.0), 0.0);
    }

    #[test]
    fn test_symlog_symmetry() {
        let pos = Scientific(1.5, 10);
        let neg = Scientific(-1.5, 10);
        assert!((pos.symlog(-50.0) + neg.symlog(-50.0)).abs() < 1e-10);
    }

    #[test]
    fn test_format() {
        assert_eq!(Scientific(0.0, 0).format(), "0");
        assert!(Scientific(1.5, 10).format().contains("e10"));
    }
}
